lib/utils/DevServerPlugin.js

```diff
@@ -0,0 +1,211 @@
+        }
+      });
+      return entriesClone;
+    };
+
+    /**
+     *
+     * Description of the option for checkInject method
+     * @typedef {Function} checkInjectOptionsParam
+     * @param {Object} _config - compilerConfig
+     * @return {Boolean}
+     */
+
+    /**
+     *
+     * @param {Boolean | checkInjectOptionsParam} option - inject(Hot|Client) it is Boolean | fn => Boolean
+     * @param {Object} _config
+     * @param {Boolean} defaultValue
+     * @return {Boolean}
+     */
+    // eslint-disable-next-line no-shadow
+    const checkInject = (option, _config, defaultValue) => {
+      if (typeof option === 'boolean') return option;
+      if (typeof option === 'function') return option(_config);
+      return defaultValue;
+    };
+
+    const compilerOptions = compiler.options;
+    compilerOptions.plugins = compilerOptions.plugins || [];
+    /** @type {Boolean} */
+    const isWebTarget = compilerOptions.externalsPresets
+      ? compilerOptions.externalsPresets.web
+      : [
+          'web',
+          'webworker',
+          'electron-renderer',
+          'node-webkit',
+          // eslint-disable-next-line no-undefined
+          undefined,
+          null,
+        ].includes(compilerOptions.target);
+    /** @type {Entry} */
+    const additionalEntries = checkInject(
+      options.injectClient,
+      compilerOptions,
+      isWebTarget
+    )
+      ? [clientEntry]
+      : [];
+
+    if (hotEntry && checkInject(options.injectHot, compilerOptions, true)) {
+      additionalEntries.push(hotEntry);
+    }
+
+    // use a hook to add entries if available
+    if (EntryPlugin) {
+      compiler.hooks.make.tapPromise('DevServerPlugin', (compilation) =>
+        Promise.all(
+          additionalEntries.map(
+            (entry) =>
+              new Promise((resolve, reject) => {
+                compilation.addEntry(
+                  compiler.context,
+                  EntryPlugin.createDependency(entry, {}),
+                  {}, // global entry
+                  (err) => {
+                    if (err) return reject(err);
+                    resolve();
+                  }
+                );
+              })
+          )
+        )
+      );
+    } else {
+      compilerOptions.entry = prependEntry(
+        compilerOptions.entry || './src',
+        additionalEntries
+      );
+      compiler.hooks.entryOption.call(
+        compilerOptions.context,
+        compilerOptions.entry
+      );
+    }
+
+    const providePlugin = new webpack.ProvidePlugin({
+      __webpack_dev_server_client__: getSocketClientPath(options),
+    });
+    providePlugin.apply(compiler);
+
+    if (
+      hotEntry &&
+      !compilerOptions.plugins.find(
+        (p) => p.constructor === webpack.HotModuleReplacementPlugin
+      )
+    ) {
+      // add and apply the HMR plugin, if it didn't exist before.
+      const plugin = new webpack.HotModuleReplacementPlugin();
+      compilerOptions.plugins.push(plugin);
```

Comment:
Just `apply` no need to push it in `plugins` array

Reply:
If it is not pushed to `plugins`, it'd add it again if the plugin is applied again.

Reply:
Can you clarify? We should not touch `compilerOptions`, only use `plugin.apply(compiler)`

Reply:
If newly applied plugins are not added to `config.plugins`, then in
```js
const compiler = webpack(config);
const server1 = new Server(compiler, options1);
server1.close(() => {
  const server2 = new Server(compiler, options2);
});
```
each time the Server constructor is called, HMR plugin and this plugin will be applied. So they will be applied twice, causing duplicate hooks and entries.

Reply:
I think using a single compiler instance across multiple dev-server instances should be reconsidered, as it's impossible to remove or update hooks and it's not possible to persist state (whether plugins and hooks are added, etc.) without touching `compiler` or `compiler.options`.

Reply:
I think here architecture problem, if we will be plugin we will not have this problem, so we need focus for v5 on refactor to plugin

Reply:
We can mark it as limitation now

Reply:
Removed `compilerOptions.plugins.push(plugin)`. This PR is ready to merge.

Reply:
The plugin would have the same limitation, since the second dev-server (`server2`) cannot get or modify the plugin added by the first dev-server (`server1`).